src/caffe/test/test_net.cpp

```diff
@@ -2224,4 +2321,52 @@ TYPED_TEST(NetTest, TestReshape) {
   }
 }
 
+TYPED_TEST(NetTest, TestSkipPropagateDown) {
+  // check bottom_need_backward if propagate_down is true
+  this->InitSkipPropNet(false);
+  vector<bool> vec_layer_need_backward = this->net_->layer_need_backward();
+  for (int layer_id = 0; layer_id < this->net_->layers().size(); ++layer_id) {
+    if (this->net_->layer_names()[layer_id] == "loss") {
+      // access to bottom_need_backward coresponding to label's blob
+      bool need_back = this->net_->bottom_need_backward()[layer_id][1];
+      // if propagate_down is true, the loss layer will try to
+      // backpropagate on labels
+      CHECK_EQ(need_back, true)
+        << "bottom_need_backward should be True";
+    }
+    if (this->net_->layer_names()[layer_id] == "ip_fake_labels")
+      CHECK_EQ(vec_layer_need_backward[layer_id], true)
+        << "layer_need_backward for ip_fake_labels should be True";
+    if (this->net_->layer_names()[layer_id] == "argmax")
+      CHECK_EQ(vec_layer_need_backward[layer_id], true)
+        << "layer_need_backward for argmax should be True";
+    if (this->net_->layer_names()[layer_id] == "innerproduct")
+      CHECK_EQ(vec_layer_need_backward[layer_id], true)
+        << "layer_need_backward for innerproduct should be True";
+  }
+  // check bottom_need_backward if propagat_down is false
+  this->InitSkipPropNet(true);
+  vec_layer_need_backward.clear();
+  vec_layer_need_backward = this->net_->layer_need_backward();
+  for (int layer_id = 0; layer_id < this->net_->layers().size(); ++layer_id) {
+    if (this->net_->layer_names()[layer_id] == "loss") {
+      // access to bottom_need_backward coresponding to label's blob
+      bool need_back = this->net_->bottom_need_backward()[layer_id][1];
+      // if propagate_down is false, the loss layer will not try to
+      // backpropagate on labels
+      CHECK_EQ(need_back, false)
+        << "bottom_need_backward should be False";
+    }
+    if (this->net_->layer_names()[layer_id] == "ip_fake_labels")
```

Comment:
and in this case, how about `if (... == "ip_fake_labels") { EXPECT_FALSE(...) } else if { ... == "argmax" } { EXPECT_FALSE(...) } else { EXPECT_TRUE(...) }` to check all the layers?